### PR TITLE
Add capability to use load directive with web urls.

### PR DESCRIPTION
**Pull request by RichiCoder1**
Work in Progress Proof of Concept for how loading scripts from the web could work.

### CODE REVIEW

**Comment by RichiCoder1** [comment]
Related to #149

**Comment by patriksvensson** [comment]
Cool! :+1: 

Just one initial thought: Maybe the load directive could download the code and keep the script processor simple?

Since the `Process` method always should process something on disc, maybe `public void Process(FilePath path, ScriptProcessorContext context)` could take something new like a `ScriptReference` that keeps an identifier (could be a path or URI or whatever) and and the path on disc? This way we could make sure that we don't import the same script twice and still be sure that we have something on disc to refer to. Would also be able to write better log messages since the directives that would instantiate those objects could add more contextual information in the future.

**Comment by RichiCoder1** [comment]
Hmm, I think that would be better. Would want to find some way to ensure we don't download the same file twice, as having a unique identifier in Process itself wouldn't help that. Maybe hash the URL and use that as the temp file's name. I just went with the most straight-forward approach first.

**Comment by RichiCoder1** [comment]
Implemented suggestion. Shares the same flaw as previously where two script references to the same script with a slightly different path string with not get rejected. I think only script file checksum will prevent that :P

File: src/Cake.Core/Scripting/ScriptReference.cs [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+using System.Security.Cryptography;
+using System.Text;
+using Cake.Core.IO;
+
+namespace Cake.Core.Scripting
+{
+    /// <summary>
+    /// Unique Script Reference
+    /// </summary>
+    public class ScriptReference
```
Comment (patriksvensson) [inline src/Cake.Core/Scripting/ScriptReference.cs]: Class should be sealed. Properties should have only getters, and their backing fields should be readonly.

File: src/Cake.Core/Scripting/ScriptReference.cs [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+            Path = localPath;
+
+            if (origin == null)
+            {
+                origin = localPath.ToString();
+            }
+            Origin = origin;
+
+            var hasher = new SHA1CryptoServiceProvider();
```
Comment (patriksvensson) [inline src/Cake.Core/Scripting/ScriptReference.cs]: is calculating a hash for the path really necessary? Wouldn't the path be enough as an ID?

File: src/Cake.Core/Scripting/Processors/LoadDirectiveProcessor.cs [hunk 3]
```diff
@@ -46,10 +48,44 @@ public override bool Process(IScriptProcessor processor, ScriptProcessorContext
...
+                var scriptPath = new FilePath(scriptPathString);
+                scriptReference = new ScriptReference(scriptPath);
+            }
+            else
+            {
+                // Download the script to a local directory.
+                FilePath tempFile = System.IO.Path.GetTempFileName();
+                try
+                {
+                    var client = new WebClient();
```
Comment (patriksvensson) [inline src/Cake.Core/Scripting/Processors/LoadDirectiveProcessor.cs]: Should make this mockable so we can write tests for this. Maybe a `IWebClient` in `Cake.Core.Net`?

File: src/Cake.Core/Scripting/ScriptReference.cs [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+
+        /// <summary>
+        /// Gets the local script path.
+        /// </summary>
+        public FilePath Path { get; private set; }
+
+        /// <summary>
+        /// Gets the script's unique token identifier;
+        /// </summary>
+        public string TokenId { get; private set; }
```
Comment (patriksvensson) [inline src/Cake.Core/Scripting/ScriptReference.cs]: I think the name `Id` would be a better name.

**Comment by patriksvensson** [comment]
Sorry for taking so long to responding to this.

I don't think it's that important to validate checksums. As long as we're not loading the same file (same path or same url) I think it will be fine. If we discover that this is something that is common for some reason, we can take a look at checksum validation.

I think your solution looks good. I've left you some comments of things I think should be addressed before merging.

File: src/Cake.Core/Scripting/Processors/LoadDirectiveProcessor.cs (hunk 3 above)
Comment (RichiCoder1) [inline src/Cake.Core/Scripting/Processors/LoadDirectiveProcessor.cs]: Hrmm. I think that might be excessive, but it would allow use to unit test this.

**Comment by RichiCoder1** [comment]
Updated with changes except WebClient request.

**Comment by devlead** [comment]
@RichiCoder1 this has unfortunately gone to far out of sync with Cake. Sorry for not getting back to you earlier.
We're closing this for now, please feel free to re-open if you have the opportunity to update.

**Pull request closed without merging**